### ISSUE TITLE
Unable to get CacheResult to work on Stateless bean in Payara 4.1.1.164 #badassfish (build 28)

### ISSUE THREAD

**Issue by mbd-dbc-dk** [payara/Payara#3052]
# JCache in Payara with Hazelcast: Cache is created but nothing ever gets cached
----------
I am trying to add caching via JSR-107 and the `@CacheResult` annotation to an existing Stateless EJB. The cache appears to be created but nothing is ever added to it => No caching happens.

## Expected Outcome

I would expect any calls to the method annotated with `@CacheResult` to be cached such that future calls could use the cached values.

## Current Outcome

For the first call to the annotated method, a cache would appear to be created. However, the cache never gets any entries, and subsequent calls to the method, using the same parameters, will still run the method. No caching occurs.

## Steps to reproduce (Only for bug reports) 

I have added `@CacheResult` to a method:

```
import javax.cache.annotation.CacheResult;

@Stateless
public class OpenOrderBusinessLogic {
.....
    @CacheResult
    public ForsRightsResponse invokeForsRights(Authentication input) {
 ....
}
```
This method is called from 

```
@WebService(...)
@Stateless
@SchemaValidation
public class OpenOrder implements OpenOrderPortType {
    ...
    
    @EJB
    private OpenOrderBusinessLogic openOrderBusinessLogic;

    ...
    
    @Override
    public CheckOrderPolicyResponse checkOrderPolicy(CheckOrderPolicyRequest input) {
        ...
        ForsRightsResponse forsRightsResponse = openOrderBusinessLogic.invokeForsRights(authentication);

```
I have added a dependency to the pom file:

```
<dependency>
            <groupId>javax.cache</groupId>
            <artifactId>cache-api</artifactId>
            <version>1.1.0</version>
</dependency>
```

This builds fine.

I have configured Payara to enable Hazelcast:

`asadmin --port 4848 --passwordfile=${PAYARA_USER_HOME}/passfile.txt set-hazelcast-configuration --enabled=true --target=server --dynamic=true
`

I have confirmed that Hazelcast is bound to CachingProvider:
```
[2018-08-16T08:43:46.907+0200] [Payara 4.1] [INFO] [] [fish.payara.nucleus.hazelcast.HazelcastCore] [tid: _ThreadID=1 _ThreadName=main] [timeMillis: 1534401826907] [levelVa
lue: 800] [[
  JSR107 Caching Provider Bound to JNDI at payara/CachingProvider]]
```

When using "FINE" for com.hazelcast in the logging, and calling the method for the first time, I get entries like this:

```
[2018-08-16T08:45:12.215+0200] [Payara 4.1] [FINEST] [] [com.hazelcast.config.Config] [tid: _ThreadID=26 _ThreadName=http-thread-pool(2)] [timeMillis: 1534401912215] [level
Value: 300] [CLASSNAME: com.hazelcast.config.Config] [[
No configuration found for dk.dbc.copa.openorder.service.OpenOrderBusinessLogic.invokeForsRights(dk.dbc.oss.ns.openorder.Authentication), using default config!]]

[2018-08-16T08:45:12.216+0200] [Payara 4.1] [FINEST] [] [com.hazelcast.config.Config] [tid: _ThreadID=82 _ThreadName=hz._hzInstance_1_development.partition-operation.thread
-1] [timeMillis: 1534401912216] [levelValue: 300] [CLASSNAME: com.hazelcast.config.Config] [[
No configuration found for dk.dbc.copa.openorder.service.OpenOrderBusinessLogic.invokeForsRights(dk.dbc.oss.ns.openorder.Authentication), using default config!]]

[2018-08-16T08:45:12.222+0200] [Payara 4.1] [INFO] [] [com.hazelcast.cache.impl.CacheService] [tid: _ThreadID=26 _ThreadName=http-thread-pool(2)] [timeMillis: 1534401912222
] [levelValue: 800] [[
[192.168.11.8]:5900 [development] [3.7.1] Added cache config: CacheConfig{name='dk.dbc.copa.openorder.service.OpenOrderBusinessLogic.invokeForsRights(dk.dbc.oss.ns.openor
der.Authentication)', managerPrefix='/hz/', inMemoryFormat=BINARY, backupCount=1, hotRestart=HotRestartConfig{enabled=false, fsync=false}}]]
```

And, the cache gets created:

```
asadmin> list-caches
Enter admin user name>  admin     
Enter admin password for user "admin"> 
{ 
HZMap payara.instance.store hz:impl:mapService 1 entries
JCache dk.dbc.copa.openorder.service.OpenOrderBusinessLogic.invokeForsRights(dk.dbc.oss.ns.openorder.Authentication) 
}

Command list-caches executed successfully.
```

However, the cache is empty:

```
asadmin> list-cache-keys --name="dk.dbc.copa.openorder.service.OpenOrderBusinessLogic.invokeForsRights(dk.dbc.oss.ns.openorder.Authentication)"
Enter admin user name>  admin
Enter admin password for user "admin"> 
{ 
}

Command list-cache-keys executed successfully.
```

And, subsequent calls to the method that should have been cached, clearly shows that the method is called again.

I have noted, however, that I get entries like these in the logfile: 

```
[#|2018-08-16T13:15:35.067+0200|WARNING|Payara 4.1|com.hazelcast.internal.cluster.impl.MulticastService|_ThreadID=114;_ThreadName=hz._hzInstance_1_development.MulticastThread;_TimeMillis=1534418135067;_LevelValue=900;|
  [192.168.11.8]:5900 [development] [3.7.1] Received data format is invalid. (An old version of Hazelcast may be running here.)
com.hazelcast.nio.serialization.HazelcastSerializationException: com.hazelcast.internal.cluster.impl.ClusterDataSerializerHook$ClusterDataSerializerFactoryImpl@46ca5df7 is not be able to create an instance for id: 34 on factoryId: 0
        at com.hazelcast.internal.serialization.impl.DataSerializableSerializer.read(DataSerializableSerializer.java:119)
        at com.hazelcast.internal.serialization.impl.DataSerializableSerializer.read(DataSerializableSerializer.java:52)
        at com.hazelcast.internal.serialization.impl.StreamSerializerAdapter.read(StreamSerializerAdapter.java:46)
        at com.hazelcast.internal.serialization.impl.AbstractSerializationService.readObject(AbstractSerializationService.java:216)
        at com.hazelcast.internal.serialization.impl.ByteArrayObjectDataInput.readObject(ByteArrayObjectDataInput.java:600)
        at com.hazelcast.internal.cluster.impl.MulticastService.receive(MulticastService.java:207)
        at com.hazelcast.internal.cluster.impl.MulticastService.run(MulticastService.java:165)
        at java.lang.Thread.run(Thread.java:748)
|#]
```
I can get rid of these by disabling multicast for Hazelcast - it does not change anything related to my caching issue though (AFAICT).

I have also tried following the instructions in #196: 

> you must copy payara-jsr107.jar from glassfish/modules to domains/domain1/lib/applibs and reference it as a library during deployment to get intereptors to work on an EJB Session Bean. This is part of the EJB spec.

This did not change anything. 

In short: It appears **the cache is being created, but not actually used**.

### Samples

Not available at this point.

## Context (Optional)

I am trying to use JCache in Payara, with "only" annotations. 

## Environment ##

- **Payara Version**: 4.1.1.164
- **Edition**: Full
- **JDK Version**: 8 u181 - Oracle
- **Operating System**: Linux 
- **Database**: NA?

**Comment by mbd-dbc-dk** [comment]
Slightly more investigation: I tried adding another Cached method, which I do not call, to the class:
/** Add a testing method - never called */
    @CacheResult
    public String foo(String input) {
        LOGGER.entry(input);
        String ret = "Never called!";
        try {
            return ret;
        } finally {
            LOGGER.exit(ret);
        }
    }

Calling my original method only, it still leads to only a single cache being created. This again leads me to believe that the cache is created by the first method call, not the first instance creation of the class.

I still have no idea why nothing is then cached.

**Comment by smillidge** [comment]
Hi @mbd-dbc-dk your organisation has a support contract. Do you want to raise a ticket there against the old version?

Otherwise 4.1.1.164 is close to two years old. Do you get the same problem on Payara 5.x?

**Comment by mbd-dbc-dk** [comment]
Hi there

Yes, I know about our support contract, but I was waiting to get the information about how to post, etc, and wanted to try my luck here.

The reason I have not tried a newer release is that this functionality is needed on a system that runs on 4.1.1.164, and also because I thought using JCache like this with Payara and Hazelcast "had to work", and that you could either point me towards my own error, confirm it "had to work", or state, "Ouch, that exact situation did not actually work in the early implementations".

The core of my "this is stumping me", is the fact that the cache is created on method call, but nothing gets entered into it (and, both argument and value is != Null). I was hoping may be that could ring a bell.

Let me discuss with the project manager, if we should investigate this on the old version or try the newest version. I don't think she will be entirely pleased with migrating the Payara version at this point, though.

**Comment by mbd-dbc-dk** [comment]
Ok, this is pretty embarrasing. The keys going into the cache was generated by JAXB, and unfortunately, they did not have a "hashCode" method. I must admit to overlooking this. Adding a hashCode method resolved my problem.

That said, the asadmin list-cache-keys command never seems to actually list anything, even for a cache with multiple entries. That is no matter to me though.
